test/src/test_derive.rs

```diff
@@ -178,6 +192,236 @@ fn test_derive_owned_to_variant() -> bool {
+        self.0
+    }
+}
+
+fn test_derive_nativeclass_omitted_inherit() -> bool {
+    println!(" -- test_derive_nativeclass_omitted_inherit");
+
+    let ok = std::panic::catch_unwind(|| {
+        let thing = Instance::<OmittedInherit, _>::new();
+        let base = thing.into_base();
+        assert_eq!(Some(54), unsafe { base.call("answer", &[]).to::<i64>() });
+    })
+    .is_ok();
+
+    if !ok {
+        godot_error!("   !! Test test_derive_nativeclass_omitted_inherit failed");
+    }
+
+    ok
+}
+
+#[derive(NativeClass)]
+#[inherit(Reference)]
+struct OmittedBasePrameter(i64);
+
+#[methods]
+impl OmittedBasePrameter {
+    fn new(_owner: &Reference) -> Self {
+        Self(54)
+    }
+
+    #[godot]
+    fn answer(&self) -> i64 {
+        self.0
+    }
+}
+
+fn test_derive_nativeclass_omitted_base_parameter() -> bool {
+    println!(" -- test_derive_nativeclass_omitted_base_parameter");
+
+    let ok = std::panic::catch_unwind(|| {
+        let thing = Instance::<OmittedBasePrameter, _>::new();
+        let base = thing.into_base();
+        assert_eq!(Some(54), unsafe { base.call("answer", &[]).to::<i64>() });
+    })
+    .is_ok();
+
+    if !ok {
+        godot_error!("   !! Test test_derive_nativeclass_omitted_base_parameter failed");
+    }
+
+    ok
+}
+
+#[derive(NativeClass)]
+#[inherit(Reference)]
+struct MethodOnBasePrameter(i64);
+
+#[methods]
+impl MethodOnBasePrameter {
+    fn new(_owner: &Reference) -> Self {
+        Self(54)
+    }
+
+    #[godot]
+    fn answer(&self, #[base] _base: &Reference) -> i64 {
+        self.0
+    }
+}
+
+fn test_derive_nativeclass_method_on_base_parameter() -> bool {
+    println!(" -- test_derive_nativeclass_method_on_base_parameter");
+
+    let ok = std::panic::catch_unwind(|| {
+        let thing = Instance::<MethodOnBasePrameter, _>::new();
+        let base = thing.into_base();
+        assert_eq!(Some(54), unsafe { base.call("answer", &[]).to::<i64>() });
+    })
+    .is_ok();
+
+    if !ok {
+        godot_error!("   !! Test test_derive_nativeclass_method_on_base_parameter failed");
+    }
+
+    ok
+}
+
+#[derive(NativeClass)]
+#[inherit(Reference)]
+struct MethodOnMacroParameters(i64);
+
+#[methods]
+impl MethodOnMacroParameters {
+    fn new(_owner: &Reference) -> Self {
+        Self(54)
+    }
+
+    #[godot(name = "ask", deref_return)]
+    fn answer(&self, #[base] _base: &Reference) -> &i64 {
+        &self.0
+    }
+}
+
+fn test_derive_nativeclass_method_on_macro_parameters() -> bool {
+    println!(" -- test_derive_nativeclass_method_on_macro_parameters");
+
+    let ok = std::panic::catch_unwind(|| {
+        let thing = Instance::<MethodOnMacroParameters, _>::new();
+        let base = thing.into_base();
+        assert_eq!(Some(54), unsafe { base.call("ask", &[]).to::<i64>() });
+    })
+    .is_ok();
+
+    if !ok {
+        godot_error!("   !! Test test_derive_nativeclass_method_on_macro_parameters failed");
+    }
+
+    ok
+}
```

Comment:
What is being tested here?

There seem to be separate tests for `name` and `deref_return`, and I wouldn't try to test attribute combinations, as that will explode exponentially. Maybe this test can be removed?

Reply:
The other test is for `#[export]`, so we are testing that `#[godot]` also accepts parameters.

But originally, it might be better to test with [trybuild](https://crates.io/crates/trybuild).

Reply:
Ah, I see. `#[export]` is deprecated though, so there's no need to add more tests. The `#[export(deref_return)]` is not even a documented feature, it just happens to work because we didn't exclude it in our implementation.

I would test `#[godot(name)]` and `#[godot(deref_return)]`, but separately. No combined tests or more `#[export]` tests.

Reply:
Since this is an integration test, we need combination tests that assumes the actual code.

Also, it helps to identify buggy code by seeing which tests are failing.

Reply:
But what is our strategy here? This will quickly lead to [combinatorial explosion](https://en.wikipedia.org/wiki/Combinatorial_explosion). So how do we decide which argument combinations we want to test?

> Also, it helps to identify buggy code by seeing which tests are failing.

I would argue testing each feature individually makes it _clearer_ which parts are broken. If the combined test fails, it can be either the of the two parts who have a bug, or only the combination of them.

Reply:
> So how do we decide which argument combinations we want to test?

Typical combinations, Combinations that may have bugs, and Most complex combinations.

For example, suppose there are individual tests `test_a` and `test_b` and a combination `test_a_and_b`.

If the test has this result, there is a bug in code tested with `test_a`:
``` 
test_a       -- fail
test_b       -- ok
test_a_and_b -- fail
```

If the test has this result, there is a bug in Higher-level code:
``` 
test_a       -- ok
test_b       -- ok
test_a_and_b -- fail
```

Reply:
Okay, let's keep the number of "combined attribute tests" small however (now is good). We can always add more tests if we see that a certain combination causes problems.

In that case, I would rename `GodotAttrOnMacroParameters` to `RenameAndDerefReturn`, and move it _after_ both `RenameMethod` and `DerefReturn` tests. It makes sense that the single attribute tests are run prior to the combinations.

Reply:
Okay, I'll  change to only fail `godot_attr_on_macro_parameters()`, so if you see that and want to remove the test, do so.